Blog: DuckDB vs ClickHouse (#16345)
PostHog/posthog.com: contents/blog/duckdb-vs-clickhouse.mdx

---
title: "In-depth: DuckDB vs ClickHouse"
date: 2026-04-14
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
author:
  - mathew-pregasen
featuredImage: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/posthog.com/contents/images/blog/posthog-engineering-blog.png
featuredImageType: full
category: Engineering
tags:
  - Explainers
  - ClickHouse
  - Data warehouse
---

import DuckDBWaitlistSurvey from 'components/DuckDBWaitlistSurvey'

DuckDB and ClickHouse are two heavyweights in the OLAP category. This makes them appear as direct competitors. In a few respects they are, but they are essentially tangential products that have garnered popularity in the OLAP world for separate reasons.

> In fact, at PostHog, we use both ClickHouse and DuckDB. Today, we want to dissect their precise differences and why an organization like us has appetite for double (OLAP) trouble.

Reviewer: ```suggestion
In fact, at PostHog, we use both ClickHouse and DuckDB. Today, we want to dissect their precise differences and discuss why an organization like us has appetite for double (OLAP) trouble.
```

Previously, I'd likened the debate between ClickHouse and Postgres to that [of grapefruits and grapes](/blog/clickhouse-vs-postgres). While the difference between ClickHouse and DuckDB isn't as drastic, it's similar to comparing onions and shallots: they share a lot in common, but are used for fundamentally different problems.

> To provide a high-level "say less" tl;dr (or whatever kids say these days): ClickHouse is a feature-packed, highly configurable database for fast analytical queries on massive data scaled across multiple instances. DuckDB, meanwhile, is the SQLite of OLAP, an embeddable and portable analytical engine that can query local or remote files (e.g. CSVs, Parquets) without any setup.

Reviewer: NOTE I don't like this, or the shallot analogy

Author: Do you not like the "fun" bits or how it positions the two tools?

Reviewer: oh ha I meant to flesh this out more before submitting

the shallot analogy just doesn't land IMO. in cooking I will sub them with few repercussions except a small bit of flavor difference.

I think the second bit I didn't like because it doesn't reflect how we are using duckdb, and I don't want people to get the wrong idea about our warehouse.

Reviewer 2: Will find a better analogy! I have very opinionated thoughts on the onion family so might be a bit biased here (though, how about onions vs spring onions?) 

Will adjust the way we talk about DuckDB

Reviewer: >  I have very opinionated thoughts on the onion family

😂 

onion vs green/spring onion (or even chives) is better!

## How we use both ClickHouse and DuckDB

> PostHog uses three databases. For PostHog's application state, we use Postgres in all of the ways you'd expect. In fact, at our start, we exclusively used Postgres until [our analytical queries were frying it alive](/blog/how-we-turned-clickhouse-into-our-eventmansion). We then implemented ClickHouse, a real-time analytical engine that powers our funnels, trends, retention, and paths over billions of events.

Reviewer: ```suggestion
PostHog uses three databases. For PostHog's application state, we use Postgres in all of the ways you'd expect. In fact, at our start, we exclusively used Postgres until [our analytical queries were frying it alive](/blog/how-we-turned-clickhouse-into-our-eventmansion). We then added ClickHouse, a real-time analytical engine that powers our funnels, trends, retention, and paths over billions of events.
```

> For a while, we were strictly a Postgres and ClickHouse stack and didn't envision implementing another OLAP solution. However, as we grew, we launched a [managed data warehouse product](/data-stack/managed-warehouse) for our customers. This was fundamentally a different product: instead of just providing analytics on customer data, we also provided a service to store and access that underlying data. Essentially, our customers could treat us as a full data stack, not just an opinionated analytics service.

Reviewer: ```suggestion
For a while, we were strictly a Postgres and ClickHouse stack and didn't envision needing another OLAP solution. However, as the needs of our customers grew, ClickHouse had constraints that meant it simply would not work for data warehousing for all of our customers. 

So, we architected a [managed data warehouse product](/data-stack/managed-warehouse) that could scale in parallel, but not tied to, our ClickHouse resources. The managed warehouse is fundamentally  different from the analytical products we offered before: instead of only providing analytics on customer data, it also provides a service to store, transform, model, and visualize _all_ your data - not just what's ingested by other PostHog products. Essentially, our customers can now use us as a full data stack, not just an opinionated analytics service.
```

> To streamline things, we built [Duckgres](https://github.com/PostHog/duckgres), a wrapper that encapsulates DuckDB in PostgreSQL wire protocol. Any service that connects to Postgres could connect to our Duckgres wrapper (and by extension, PostHog's managed DuckDB instance). It's kinda neat: our customers could connect any Postgres-compatible [BI tool](/data-stack/business-intelligence), [dbt](/data-stack/data-modeling), or Postgres service to the DuckDB warehouse without knowing it's DuckDB.

Reviewer: ```suggestion
To fit into the existing modern data stack ecosystem, we built [Duckgres](https://github.com/PostHog/duckgres), a wrapper that encapsulates DuckDB in PostgreSQL wire protocol. Any service that connects to Postgres could connect to our Duckgres wrapper (and by extension, PostHog's managed DuckDB instance). It's kinda neat: our customers could connect any Postgres-compatible [BI tool](/data-stack/business-intelligence), [dbt](/data-stack/data-modeling), or Postgres service to the DuckDB warehouse without worrying about DuckDB-specific compatibility. Nearly every tool out there connects to Postgres – and by extension – now can connect to our managed warehouses.
```

> ## ClickHouse is massive (literally)

Reviewer: This feels abrupt. We were doing story time and now we're at "because clickhouse is modeled closer...". I feel like it needs a transition. like "so what's the difference"

Author: Yeah, this is a good point. I think we move the "How we use both ClickHouse and DuckDB" section to the bottom and combine it with the conclusion.

> Because ClickHouse is modeled closer to the median OLAP/columnar database, it's best to start with it. ClickHouse is a dedicated server for handling and returning analytical data. It is a long-lived process that scales horizontally across nodes with self-managed storage. It is *incredibly* fast, made possible by its materialized views, mature compression, and sparse primary indexes. It's fantastic at parallelization and is designed for scaling analytical data across multiple servers.

Reviewer: idk if you can really say "it's best to start with it". For who to start with it, and for what purpose? For finely tuned analytics - yes best to start with CH. But for adhoc queries, CH isn't your best bet.

Reviewer 2: Ah apologies, I meant it as starting with covering ClickHouse, then DuckDB (as in article ordering). Will adjust so it's more clear.

![ClickHouse scaling](https://res.cloudinary.com/dmukukwp6/image/upload/w_1600,c_limit,q_auto,f_auto/Click_House_3b27ea474d.png)

DuckDB couldn't be more different. DuckDB is just a single process that scales vertically, all embedded in a single file. When you run out of resources, you'll need to bump up the instance or migrate DuckDB to a beefier server.

![DuckDB scaling](https://res.cloudinary.com/dmukukwp6/image/upload/w_1600,c_limit,q_auto,f_auto/Duck_DB_3a01cf6c0a.png)

This also means that DuckDB evades the complexities of horizontally scaling. ClickHouse (via Apache ZooKeeper or ClickHouse Keeper) has to track every data part, every merge, and every insert across every replica. That's a challenge when dealing with thousands of tables and millions of ingested events. To address this, some developers prefer ClickHouse's paid offering, ClickHouse Cloud which stores data in object storage (e.g. S3), treating compute nodes as stateless workers with a cache.

> The main challenge with scaling DuckDB is the compute-storage coupling issue, where pre-spec'd servers might only ship with compute and storage coupled, leading to one of the two resources being over-provisioned.

Reviewer: so as far as I'm asware clickhouse is a challenge because storage and compute are tied, and that's not the case with how we are using duckdb. with our implementation each duck process gets a box but storage is completely separate because we are using ducklack as the storage format with postgres catalog so we can horizontally scale compute without impacting storage

## DuckDB is lightweight

DuckDB is not a server. Instead, it functions like SQLite: an embedded, in-process engine that runs as an application. This makes DuckDB easy to spin up, query on local data (including Parquet, CSV, and JSON files) or data in a storage bucket, and spin down. It can also be pointed to a remote bucket and operate the same. It's ephemeral, single-user, and saves metadata to a local and portable `.duckdb` database file (or an external file it's pointed to).

![DuckDB query](https://res.cloudinary.com/dmukukwp6/image/upload/w_1600,c_limit,q_auto,f_auto/Duck_DB_q_79607d6f3c.png)

ClickHouse is typically faster than DuckDB at heavy queries given its massive toolset to optimize retrieval and pre-calculate aggregates; however, DuckDB is fantastic at spontaneous, smaller analytical queries as it skips client-server communication and operates on local data in a stock setup. Data is ingested into DuckDB in batches. For many users, data is just ingested once, queried, and then the DuckDB instance is destroyed.

> 

Author: Maybe we put a bit more here about DuckDB as a query engine vs DuckDB as a database. DuckDB as a query engine feels connected to clickhouse-local while the infrastructure comparison feels more about DuckDB as a database.

![DuckDB scaling](https://res.cloudinary.com/dmukukwp6/image/upload/w_1600,c_limit,q_auto,f_auto/Duck_DB_3a01cf6c0a.png)

This also means that DuckDB evades the complexities of horizontally scaling. ClickHouse (via Apache ZooKeeper or ClickHouse Keeper) has to track every data part, every merge, and every insert across every replica. That's a challenge when dealing with thousands of tables and millions of ingested events. To address this, some developers prefer ClickHouse's paid offering, ClickHouse Cloud which stores data in object storage (e.g. S3), treating compute nodes as stateless workers with a cache.

The main challenge with scaling DuckDB is the compute-storage coupling issue, where pre-spec'd servers might only ship with compute and storage coupled, leading to one of the two resources being over-provisioned.

## Interface match-up: connecting to ClickHouse versus DuckDB

A second point of difference is how you interface with ClickHouse and DuckDB.

> Because DuckDB is embedded, it is queried on-device. Of course, it might be accessed via an SSL tunnel or a proxied service (e.g. Duckgres), but its base implementation involves no server-client relationship. Just an `import` statement, a SQL query, and that's it. It reads and writes local files natively, scans remote Parquet on S3, and integrates tightly with Pandas/Polars DataFrames. It only supports a single writer at a time (but can support concurrent reads and is multithreaded).

Author: Can we say "DuckDB query engine" somewhere in here.

Reviewer 3: i believe the added callout addresses this

Despite their interface differences, both databases speak a SQL variant. The ClickHouse variant has some non-standard syntax like array functions, the `FINAL` keyword for deduplication, etc. DuckDB's variant is closer to stock Postgres, but with some nice additions like `EXCLUDE`, `COLUMNS(*)` regex matching, and direct file-path-as-table-name queries.

## Performance match-up: measuring ClickHouse and DuckDB

Comparing ClickHouse and DuckDB on performance is a bit tricky. On small-to-moderate datasets, DuckDB can match or beat ClickHouse on cold queries since it avoids network overhead and has an efficient vectorized engine. But on large datasets with tuned materialized views and indexes, ClickHouse will always pull ahead.

This is a bit of a flawed comparison, however. The purpose of ClickHouse is to serve long-standing queries. It's a heavily tunable database. For example, we'll have a ClickHouse query that aggregates all of the views of the database to establish the average time spent per session – it's the same query that could be pre-empted by materialized views. DuckDB, meanwhile, is designed to serve ad hoc queries at blazing-fast speed with no set-up.

> 

Author: Maybe we have another paragraph here talking about how it's possible for DuckDB to scale? Just using it as query engine + having more scalable formats backing it?

## Notable ClickHouse and DuckDB libraries

There are two other ClickHouse and DuckDB libraries that lend to their flexibility: clickhouse-local and DuckLake. To provide a holistic picture of both ecosystems, we'd like to acknowledge these.

### What is clickhouse-local?

Just like how DuckDB has a lightweight (query engine) and heavyweight (database) implementation, ClickHouse also has variations.

[clickhouse-local](https://clickhouse.com/docs/operations/utilities/clickhouse-local) is a portable distribution of ClickHouse that's embedded without a full database server. In many ways, it's the DuckDB version of ClickHouse, as it can index Parquet files and has been [benchmarked against DuckDB](https://www.vantage.sh/blog/clickhouse-local-vs-duckdb).

However, by ClickHouse's own position, clickhouse-local was not designed for serving end users or applications. It's more of a portable tool for testing queries and one-off data processing than something designed to serve production applications or scale horizontally. So, for the rest of this piece, we'll stick to the production-grade ClickHouse build.

### What is DuckLake?

> DuckLake is a data lake by the DuckDB team. For those unfamiliar, data lakes store raw, unstructured data side-by-side with structured data for future querying. DuckLake provides this flexibility with a native integration with DuckDB. Unlike DuckDB, DuckLake isn't subject to the same single machine sizing constraints as DuckDB and can be used for durable, massive storage for datasets intended for DuckDB queries.

Reviewer 4: ```suggestion
DuckLake is a data lake built by the DuckDB team. For those unfamiliar, data lakes store raw, unstructured data side-by-side with structured data for future querying. DuckLake provides this flexibility with a native integration with DuckDB. Unlike DuckDB, DuckLake isn't subject to the same single machine sizing constraints as DuckDB and can be used for durable, massive storage for datasets intended for DuckDB queries.
```
